### PR TITLE
Use codespan for analyzer errors

### LINKED ISSUE #382: expression is not an identifier at compiler/src/yul/mappers/assignments.rs:77:9
Compiling this file:

```
contract o:
 pub def r():
  6=0
```

```
thread 'main' panicked at 'expression is not an identifier', compiler/src/yul/mappers/assignments.rs:77:9
stack backtrace:
   0:     0x55c3cb5adc20 - std::backtrace_rs::backtrace::libunwind::trace::h9d49145f95eb5894
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x55c3cb5adc20 - std::backtrace_rs::backtrace::trace_unsynchronized::hab1d020365bb6864
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55c3cb5adc20 - std::sys_common::backtrace::_print_fmt::h7659588431e304bd
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x55c3cb5adc20 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h09f4a9e3befae3c7
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x55c3cb5c809c - core::fmt::write::hf3fdfde304b9a088
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/core/src/fmt/mod.rs:1092:17
   5:     0x55c3cb5aab82 - std::io::Write::write_fmt::h1cb850689c7116f0
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/io/mod.rs:1568:15
   6:     0x55c3cb5afe85 - std::sys_common::backtrace::_print::hdbccd5aa093ba544
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x55c3cb5afe85 - std::sys_common::backtrace::print::hc639c4f320222558
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x55c3cb5afe85 - std::panicking::default_hook::{{closure}}::hdb012dd7a485bb5d
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/panicking.rs:208:50
   9:     0x55c3cb5af933 - std::panicking::default_hook::h75facbce77b6ba91
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/panicking.rs:225:9
  10:     0x55c3ca8854df - fe_compiler::errors::DEFAULT_PANIC_HOOK::{{closure}}::{{closure}}::h6ae12772f900e367
  11:     0x55c3cb5b0646 - std::panicking::rust_panic_with_hook::hbcaa5de2cb5e22d5
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/panicking.rs:595:17
  12:     0x55c3ca97adf3 - std::panicking::begin_panic::{{closure}}::h0fd99842fee6f975
  13:     0x55c3ca97adbe - std::sys_common::backtrace::__rust_end_short_backtrace::h204cc64e10aa1ad2
  14:     0x55c3ca953e0c - std::panicking::begin_panic::h7e1ce5a560904b72
  15:     0x55c3ca96ea0b - fe_compiler::yul::mappers::assignments::assign::hf28059932c72e7e6
  16:     0x55c3ca8e4fb0 - fe_compiler::yul::mappers::functions::func_stmt::h50fcd4f4a120751f
  17:     0x55c3ca8b64bc - <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold::h7684b63c719d7c26
  18:     0x55c3ca89e0b0 - <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter::h545531e71c39b1aa
  19:     0x55c3ca8e3e2d - fe_compiler::yul::mappers::functions::func_def::hbace2e55acfc4f32
  20:     0x55c3ca97b804 - fe_compiler::yul::mappers::contracts::contract_def::h09ca86dfdba91ca2
  21:     0x55c3ca9544e3 - fe_compiler::yul::mappers::module::module::hb739e3545351a339
  22:     0x55c3ca903e49 - fe_compiler::yul::compile::hc99409b781963933
  23:     0x55c3ca8f5577 - fe_compiler::compile::h48b563e45d620494
  24:     0x55c3ca872eb5 - fe::main::hf599b79e8faf1ec5
  25:     0x55c3ca8798f6 - std::sys_common::backtrace::__rust_begin_short_backtrace::hacda71462a60a3cf
  26:     0x55c3ca878378 - std::rt::lang_start::{{closure}}::hab486fd58046af73
  27:     0x55c3cb5b0a3a - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h31ee16c075ae2d62
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/core/src/ops/function.rs:259:13
  28:     0x55c3cb5b0a3a - std::panicking::try::do_call::h3993344bf97066b6
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/panicking.rs:379:40
  29:     0x55c3cb5b0a3a - std::panicking::try::h09e523df7c9c2d28
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/panicking.rs:343:19
  30:     0x55c3cb5b0a3a - std::panic::catch_unwind::h9d4ddb10ebb815ff
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/panic.rs:431:14
  31:     0x55c3cb5b0a3a - std::rt::lang_start_internal::hc92e27a69d75de2a
                               at /rustc/9c09c1f7cfcf9de0522bcd1cfda32b552195c464/library/std/src/rt.rs:51:25
  32:     0x55c3ca87834b - std::rt::lang_start::hfed311bddf8839f0
  33:     0x7ff106ca0bf7 - __libc_start_main
  34:     0x55c3ca86fbea - _start
  35:                0x0 - <unknown>

You've hit an internal compiler error. This is a bug in the Fe compiler.
Fe is still under heavy development, and isn't yet ready for production use.

If you would, please report this bug at the following URL:
  https://github.com/ethereum/fe/issues/new
```

On latest github master, built with

```
[profile.dev]
debug = true
panic = "abort"
```

for fuzzing.  Using https://github.com/agroce/afl-compiler-fuzzer

### LINKED ISSUE #408: More descriptive error message for CannotMove error
### What is wrong?

The following snippet returns `self.messages[addr]` without calling `to_mem()`

```
contract GuestBook:
  messages: map<address, bytes[100]>

  pub def get_msg(addr: address) -> bytes[100]:
      return self.messages[addr]
```

The error message isn't very user friendly.

```
Unable to compile guest_book.fe.
Analyzer error: CannotMove on line 8
pub def get_msg(addr: address) -> bytes[100]:
      return self.messages[addr]
```

### How can it be fixed

The error message should explain and hint towards `to_mem()` possibly even linking to http://fe.ethereum.org/docs/spec/index.html#623-the-to_mem-function

### PR DESCRIPTION
**Pull request by sbillig**
### What was wrong?

Analyzer errors weren't fancy enough.

### How was it fixed?

This has become a behemoth of a pr, sorry for the hassle.

The context object now contains a vec of diagnostics. Analysis functions will return `Ok` with correct type information wherever possible, even if they emit an error diagnostic. In cases where correct type information can't be returned, `Err(SemanticError::fatal())` is returned. There are still a lot of old errors yet to be converted. In some cases, fancy errors will be followed by an old error, so the user experience will be a little awkward for a while. When they've all been converted to codespan diagnostics, SemanticError/ErrorKind can finally be retired.

I had to do a lot of refactoring to give myself enough context to emit pretty error messages. I've been inlining some utility functions and passing around more nodes and spans, and I'm starting to wonder if there's a better way. Eg rustc's "hir" analysis uses various query functions to retrieve such info. Maybe we could structure things similarly as part of the salsa experiment and avoid passing around spans everywhere. Either way I need to take a breather from this for a while; maybe some enlightenment will come later. Y'all are more than welcome to pick away at the remaining old errors, and/or suggest better wording for the errors I did convert. The quality of the error messages is very reflective of my energy level at the time they were written :)

I added a basic check of the left-hand side of assignments; this could be extended to verify that the lhs isn't a read-only attribute of a built-in, for example. The labels of event and struct constructor args are now verified as well. The labels of function args are still ignored (I'd like extend the same checking logic of events/structs to function calls, if the ideas in #397 are agreed upon).

Along the way, I noticed that tuples inside maps weren't being lowered and would thus panic; couldn't resist fixing that, even though it could have been a separate pr.

closes #408
closes #382

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/spec/index.md) if applicable
- [x] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

File: analyzer/src/traversal/assignments.rs [hunk 1]
```diff
@@ -39,6 +66,30 @@ pub fn assign(
     unreachable!()
 }
 
+pub fn check_assign_target(
```
Comment (sbillig) [inline analyzer/src/traversal/assignments.rs:69]: I now notice that I didn't apply the same check to aug_assign. That can wait for a future pr.

**Comment by sbillig** [comment]
@satyamakgec I'd be happy to hold off on this until #385 is merged and deal with the merge conflicts, if your stuff is nearly ready.

**Comment by satyamakgec** [comment]
Thanks @sbillig, Yeah it is ready for review once it is merged then you will get a lot of conflicts

File: analyzer/src/traversal/expressions.rs [hunk 2]
```diff
@@ -529,20 +579,162 @@ fn expr_call_builtin_function(
...
+
+pub fn validate_arg_types(
+    scope: Shared<BlockScope>,
+    context: Shared<Context>,
+    name: &str,
+    args: &Node<Vec<Node<fe::CallArg>>>,
+    params: &[(String, FixedSize)],
+) -> Result<(), SemanticError> {
+    for ((label, param_type), arg) in params.iter().zip(args.kind.iter()) {
+        let val_attrs = assignable_expr(Rc::clone(&scope), Rc::clone(&context), &arg.kind.value)?;
```
Comment (sbillig) [inline analyzer/src/traversal/expressions.rs:689]: I just noticed that if there are more args than the fn expects, the excess args won't be analyzed. I'm inclined to fix this later.

**Comment by g-r-a-n-t** [comment]
:eye:

File: analyzer/src/traversal/expressions.rs (hunk 2 above)
Comment (g-r-a-n-t) [reply analyzer/src/traversal/expressions.rs:689]: not a big deal

**Review by g-r-a-n-t — APPROVED** [review]
Looks great. Alot of nice improvements in here.

**Pull request merged**